### PR TITLE
internal/router: add support for id resolving

### PR DESCRIPTION
**Pull request by rogpeppe**
We allow non-specific ids in API paths.
The details of the id resolution are kept
out of the router - that responsibility is
up to the API version implemention,
and will be implemented in a subsequent branch.

### CODE REVIEW

File: internal/router/router_test.go [hunk 1]
```diff
@@ -255,12 +269,39 @@ var routerTests = []struct {
...
+			url.Series = series
+		}
+		if url.Revision == -1 {
+			url.Revision = revision
+		}
+		return nil
+	}
+}
+
+func resolveURLError(*charm.URL) error {
```
Comment (frankban) [inline internal/router/router_test.go:300]: Is this used? Do we need to add a test for resolveUrl error?

File: internal/router/router_test.go [hunk 2]
```diff
@@ -255,12 +269,39 @@ var routerTests = []struct {
 	},
 }}
 
+// resolveURL returns a URL resolver that resolves
+// unspecified series and revision to the given series
+// and revision.
+func resolveURL(series string, revision int) func(*charm.URL) error {
```
Comment (frankban) [inline internal/router/router_test.go]: s/resolveURL/makeResolveURL/ ?

**Comment by frankban** [comment]
:+1:  with possible minor changes.
Thank you!

File: internal/router/router_test.go (hunk 1 above)
Comment (rogpeppe) [inline internal/router/router_test.go:300]: good point.

File: internal/router/router_test.go (hunk 2 above)
Comment (rogpeppe) [inline internal/router/router_test.go]: done

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**